rfcs/primitive_lists_api.md

```diff
@@ -0,0 +1,283 @@
+
+beginsWith(String)
+beginsWith(String, Case)
+between(Date, Date)
+between(long, long)
+between(float, float)
+between(double, double)
+contains(String)
+contains(String, Case)
+distinct()
+endsWith(String)
+endsWith(String, Case)
+equalTo(Boolean)
+equalTo(Byte)
+equalTo(byte[])
+equalTo(Date)
+equalTo(Double)
+equalTo(Float)
+equalTo(Integer)
+equalTo(Long)
+equalTo(Short)
+equalTo(String)
+equalTo(String, Case)
+findAllSorted()
+findAllSortedAsync()
+greaterThan(Date)
+greaterThan(double)
+greaterThan(float)
+greaterThan(long)
+greaterThanOrEqualTo(Date)
+greaterThanOrEqualTo(double)
+greaterThanOrEqualTo(float)
+greaterThanOrEqualTo(long)
+in(Boolean[])
+in(Byte[])
+in(Date[])
+in(Double[])
+in(Float[])
+in(Integer[])
+in(Long[])
+in(Short[])
+in(String[])
+in(String[], Case)
+isEmpty()
+isNotEmpty()
+isNotNull()
+isNull()
+lessThan(Date)
+lessThan(double)
+lessThan(float)
+lessThan(long)
+lessThanOrEqualTo(Date)
+lessThanOrEqualTo(double)
+lessThanOrEqualTo(float)
+lessThanOrEqualTo(long)
+like(String)
+max()
+maximumDate()
+min()
+minimumDate()
+notEqualTo(Boolean)
+notEqualTo(Byte)
+notEqualTo(byte[])
+notEqualTo(Date)
+notEqualTo(Double)
+notEqualTo(Float)
+notEqualTo(Integer)
+notEqualTo(Long)
+notEqualTo(Short)
+notEqualTo(String)
+notEqualTo(String, Case)
+sum()
+```
+
+Benefit: Intent expressed in method params. All queries use the same class Downside: Ton of new methods. More chances for
+people to use the "wrong" method. Today people still have to remember not to call e.g. `sum` on a String field though.
+
+
+4b) Add support for a special keyword for column names. NSPredicate apparently uses `SELF`, `$` could also work.
+
+Benefit: No need to add a lot of new methods. Easier to upgrade from existing workaround where people reference
+a field inside a `RealmInt` type (search-replace-done). Downside: Would not work if people used the restricted name for
+an actual field name. Syntax would also be a bit awkward and "hidden".
+
+
+# Design 2: PrimitiveRealmList<Object>
+
+In this variant, we add 3 new classes `PrimitiveRealmList`, `PrimitiveRealmQuery`, `PrimitiveRealmResults`
+
+1a) Add new classes
+
+```
+// New classes for supporting primitive arrays
+public class PrimitiveRealmList<Object> extends AbstractList<E> implements OrderedRealmCollection<E>
+public class PrimitiveRealmQuery<Object>
+public class RealmResults<Object> implements OrderedRealmCollection<E>
+
+// We need to relax the Collection Interface methods unless we want to create a completely new Collection hierarchy for
+// primitive types.
+public interface OrderedRealmCollection<E> extends List<E>, RealmCollection<E>
+public interface RealmCollection<E> extends Collection<E>
+```
+
+1b) Add new classes with `Primitive*` as an abstract super class:
+
+Precedence: SparseIntArray, SparseBooleanArray, etc.
+
+```
+protected abstract class PrimitiveRealmQuery<Object>
+public class PrimitiveIntegerRealmQuery extends PrimitiveRealmQuery<Integer>
```

Comment:
Shouldn't we have `public class PrimitiveLongRealmQuery extends PrimitiveRealmQuery<Long>`?